# <font color='blue'> Module 2</font>

## Practice Numpy and Pandas with datasets

In [1]:
# generating the datasets

import pandas as pd
import numpy as np

# Seed for reproducible randomness
np.random.seed(42)
n_records = 1000

# Continuous raw data setup
years_list = list(range(2020, 2026)) # 2020 to 2025
raw_sales = np.random.randint(5000, 75000, size=(n_records, len(years_list)))

data = {
    'UniqID': [f'TRX-{1000 + i}' for i in range(n_records)],
    'ProdGroup': np.random.choice(['Electronics', 'Apparel', 'Home & Kitchen', 'Beauty', 'Sports'], n_records),
    'MarketSegment': np.random.choice(['North America', 'Europe', 'Asia-Pacific', 'Latin America', 'MEA'], n_records),
    'Fulfillment_Type': np.random.choice(['Express', 'Standard', 'Economy'], n_records),
    'Quality_Score_Avg': np.random.uniform(3.5, 5.0, n_records).round(1),
    'Tax_Exempt_Status': np.random.choice([True, False], n_records, p=[0.15, 0.85])
}

# Integrate multi-year continuous matrices
for idx, yr in enumerate(years_list):
    data[yr] = raw_sales[:, idx]

df_sales = pd.DataFrame(data)
print("E-Commerce Dataset with 1,000 records successfully loaded!")
print(f"Shape: {df_sales.shape}")

E-Commerce Dataset with 1,000 records successfully loaded!
Shape: (1000, 12)


### <font color='blue'>Exercise 1: Structural Dropping & Quick Assessment</font>

1. **Inspect the top 3 rows** of `df_sales`.
2. **Inspect the bottom 3 rows**.
3. **Drop the columns** `UniqID`, `Fulfillment_Type`, and `Tax_Exempt_Status` along the correct axis in-place.
4. **Print the remaining shape** of the DataFrame.

In [6]:
#  1. Inspect the top 3 rows of df_sales
df_sales.head(3)

,UniqID,ProdGroup,MarketSegment,Fulfillment_Type,Quality_Score_Avg,Tax_Exempt_Status,2020,2021,2022,2023,2024,2025
0,TRX-1000,Home & Kitchen,North America,Economy,3.6,False,20795,5860,59886,11265,42194,49131
1,TRX-1001,Sports,North America,Standard,3.8,False,65263,21023,46090,72221,69820,5769
2,TRX-1002,Beauty,Europe,Express,3.5,False,64735,67955,69925,72969,10311,58707


In [9]:
# 2. Inspect the bottom 3 rows
df_sales.tail(3)

,UniqID,ProdGroup,MarketSegment,Fulfillment_Type,Quality_Score_Avg,Tax_Exempt_Status,2020,2021,2022,2023,2024,2025
997,TRX-1997,Beauty,Europe,Economy,3.6,False,14079,13399,65726,21532,48162,8749
998,TRX-1998,Beauty,North America,Economy,4.4,False,16353,33477,29570,16437,43104,68370
999,TRX-1999,Beauty,North America,Economy,4.4,False,9551,73359,59880,17916,46445,52636


In [12]:
# 3. Drop the columns UniqID, Fulfillment_Type, and Tax_Exempt_Status along the correct axis in-place
df_subset_sales = df_sales.drop(['UniqID', 'Fulfillment_Type', 'Tax_Exempt_Status'], axis=1)
df_subset_sales.head(3)

,ProdGroup,MarketSegment,Quality_Score_Avg,2020,2021,2022,2023,2024,2025
0,Home & Kitchen,North America,3.6,20795,5860,59886,11265,42194,49131
1,Sports,North America,3.8,65263,21023,46090,72221,69820,5769
2,Beauty,Europe,3.5,64735,67955,69925,72969,10311,58707


In [15]:
# 4. Print the remaining shape of the DataFrame.
print(df_subset_sales.shape)

(1000, 9)


## <font color='blue'>Exercise 2: Column Renaming & Feature Generation</font>

1. **Rename the metadata columns** so they read cleanly: Change `ProdGroup` to `Category` and `MarketSegment` to `Region`.
2. **Generate a new metric column** named `Total_Revenue` that sums up all values horizontally across all year periods for each distinct index row profile.

In [18]:
# 1. Rename the metadata columns so they read cleanly: Change 'ProdGroup' to 'Category' and 'MarketSegment' to 'Region'

df_cleaned = df_subset_sales.rename(columns={'ProdGroup':'Category', 'MarketSegment':'Region'})
df_cleaned.head(3)

,Category,Region,Quality_Score_Avg,2020,2021,2022,2023,2024,2025
0,Home & Kitchen,North America,3.6,20795,5860,59886,11265,42194,49131
1,Sports,North America,3.8,65263,21023,46090,72221,69820,5769
2,Beauty,Europe,3.5,64735,67955,69925,72969,10311,58707


In [23]:
# 2. Generate a new metric column named 'Total_Revenue' that sums up all values horizontally across all year periods 
# for each distinct index row profile

years=[2020, 2021, 2022, 2023, 2024, 2025]

df_cleaned['Total_Revenue'] = df_cleaned[years].sum(axis=1)
df_cleaned.head(3)

,Category,Region,Quality_Score_Avg,2020,2021,2022,2023,2024,2025,Total_Revenue
0,Home & Kitchen,North America,3.6,20795,5860,59886,11265,42194,49131,189131
1,Sports,North America,3.8,65263,21023,46090,72221,69820,5769,280186
2,Beauty,Europe,3.5,64735,67955,69925,72969,10311,58707,344602


## <font color='blue'>Exercise 3: Label Re-indexing</font>

Currently, our index is a meaningless structural range number sequence (0 to 999).

1. **Re-index the entire DataFrame permanently** using the `Category` data dimension column.
2. **Filter the explicit index label location** to pull every single matching record segment corresponding exactly to `Electronics`.

In [24]:
# 1. Re-index the entire DataFrame permanently using the 'Category' data dimension column
df_cleaned.set_index('Category', inplace=True)
df_cleaned.head(3)

,Region,Quality_Score_Avg,2020,2021,2022,2023,2024,2025,Total_Revenue
Category,,,,,,,,,
Home & Kitchen,North America,3.6,20795,5860,59886,11265,42194,49131,189131
Sports,North America,3.8,65263,21023,46090,72221,69820,5769,280186
Beauty,Europe,3.5,64735,67955,69925,72969,10311,58707,344602


In [32]:
# 2. Filter the explicit index label location to pull every single matching record segment corresponding exactly to 'Electronics'
print(df_cleaned.loc['Electronics'])

                    Region  Quality_Score_Avg   2020   2021   2022   2023  \
Category                                                                    
Electronics         Europe                5.0  74092   8890  46606  15627   
Electronics  Latin America                4.0  72121  74479  24457  71557   
Electronics  North America                4.3  68335  15965  29538  13110   
Electronics         Europe                3.6  74163  30939  53925  47941   
Electronics  North America                3.6  31105  20707  26976  49262   
...                    ...                ...    ...    ...    ...    ...   
Electronics   Asia-Pacific                4.4  10666  42122  30092  31505   
Electronics         Europe                4.9  48169  56742  45522   8880   
Electronics  North America                4.6  41701   8385  62241  15367   
Electronics  Latin America                4.9  73653  41929  53291   9042   
Electronics         Europe                4.2  24950  20173  63032  33036   

## <font color='blue'>Exercise 4: Explicity Label Data Typing & Slicing Variable Generation</font>

1. **Safely transform every column header item** mapping inside your list configuration profile into explicit Python string formats (`str`).
2. **Generate a standalone list variable container** called `revenue_years` capturing historical string years from `'2020'` through to `'2025'`.
3. **Use that explicit variable list container map** to pull back a localized data view segment strictly tracking `Apparel` performance metrics during those targeted years.

In [37]:
# 1. Safely transform every column header item mapping inside your list configuration profile into explicit Python string formats (str)

df_cleaned.columns = list(map(str, df_cleaned.columns))
df_cleaned.head(3)

,Region,Quality_Score_Avg,2020,2021,2022,2023,2024,2025,Total_Revenue
Category,,,,,,,,,
Home & Kitchen,North America,3.6,20795,5860,59886,11265,42194,49131,189131
Sports,North America,3.8,65263,21023,46090,72221,69820,5769,280186
Beauty,Europe,3.5,64735,67955,69925,72969,10311,58707,344602


In [38]:
# 2. Generate a standalone list variable container called revenue_years capturing historical string years from '2020' through to '2025'

revenue_years = list(map(str, range(2020, 2026)))
print(revenue_years)

['2020', '2021', '2022', '2023', '2024', '2025']


In [40]:
# 3. Use that explicit variable list container map to pull back a localized data view segment strictly tracking 'Apparel' 
# performance metrics during those targeted years

apparel_history = df_cleaned.loc['Apparel', revenue_years]
print(apparel_history.head())

           2020   2021   2022   2023   2024   2025
Category                                          
Apparel   33693  30658  23431   7747  64150  70725
Apparel   40773  72435  61886  71803  36551  16394
Apparel   14692  50758  70697  42065  37606  16534
Apparel   45397   6016  60591  28247  29300  14268
Apparel    5854  43623  12392  60680  51717  55859


## <font color='blue'> Exercise 5: Multi-Criteria Intersection Filtering</font>

Find out how many operations meet this intersection criteria:

1. **The transaction record belongs** to the `'Asia-Pacific'` region.
2. **AND the metric trend tracking data row registers** a `'Quality_Score_Avg'` score greater than or equal to `4.5`.

In [50]:
# 1. The transaction record belongs to the 'Asia-Pacific' region

cond_asia_pacific = (df_cleaned['Region'] == 'Asia-Pacific')
df_asia_pacific = df_cleaned[cond_asia_pacific]
cond_asia_pacific.head()

Category
Home & Kitchen    False
Sports            False
Beauty            False
Apparel           False
Apparel           False
Name: Region, dtype: bool

In [51]:
# 1+2

cond_asia_pacific_and = (df_cleaned['Region'] == 'Asia-Pacific') & (df_sales['Quality_Score_Avg'] >= 4.5)
df_asia_pacific = df_cleaned[cond_asia_pacific]
cond_asia_pacific_and.head()

Category
0    False
1    False
2    False
3    False
4    False
dtype: bool